Exploration 7.2

\textbf{2}
We write an euler code to approximate the solution at $y(10)$
_____

In [ ]:
def f(t,y):
    return np.exp(t)*np.sin(y)

In [ ]:
def soln(t):
    return 2 * np.arctan(1 / (17.9858 * np.exp(-np.exp(t))))

In [85]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.30
h = 0.30
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: 10909.871541508808


In [86]:
analytical_prediction= soln(10)

ratio= y_10/analytical_prediction
print("Analytical prediction of y(10) is", analytical_prediction)
print("Ratio of numerical vs analytical is", ratio)


Analytical prediction of y(10) is 3.141592653589793
Ratio of numerical vs analytical is 3472.719968657446


C:\Users\Hasif\AppData\Local\Temp\ipykernel_31412\1519247141.py:2: RuntimeWarning: divide by zero encountered in double_scalars
  return 2 * np.arctan(1 / (17.9858 * np.exp(-np.exp(t))))


We see from the printed ratio, the numerical approximation blow up as it is nowhere close to analytical solution of $y(10)$

_____


3
____

We now vary our stepsizes.

In [90]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.001
h = 0.001 
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: -1179.6728318223827


In [93]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.002
h = 0.002
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: 522.1741061875068


In [95]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.003
h = 0.003
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: -462.84192139051936


We see that for little perturbations to initial conditions the y(10) varies drastically-- For 0.001, the prediction is -1179.6728318223827 and For stepsize 0.002, the prediction is 522.1741061875068 and for stepsize 0.003 the prediction is -462.84192139051936. This shows the sensitivity of initial condition variance. This portrayal fulfills step 3. 

_____
4
Now we perturb our initial conditions.

In [96]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.301
h = 0.30
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: -683.260730706581


In [97]:
import math

def euler_method(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        y += h * math.exp(t) * math.sin(y)
        t += h
    return y

# Given values
t0 = 0
y0 = 0.302
h = 0.30
t_end = 10

y_10 = euler_method(t0, y0, h, t_end)
print("Approximate value of y(10) using Euler's method:", y_10)

Approximate value of y(10) using Euler's method: 12467.542923116842


Notice how varying initial conditions from y(0)= 0.30 to y(0)=0.31, 0.32 changes our numerical approximation drastically. Given subtle perturbations to initial conditions bring these drastic changes we can claim that it exhibits sensitive dependence on initial conditions
____

We now throw our hand at using Runge-Kutta method for the numerical approximation to satisfy no.6

6
______


![Alt text](image.png)

In [98]:
import math

def f(t, y):
    return math.exp(t) * math.sin(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 0.3
h = 0.3
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)


Approximate value of y(10) using 4th order Runge-Kutta method: 13386.789636711563


Now, let's vary stepsizes

In [101]:
import math

def f(t, y):
    return math.exp(t) * math.sin(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 0.30
h = 0.001
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)


Approximate value of y(10) using 4th order Runge-Kutta method: 830.3944247777524


In [102]:
import math

def f(t, y):
    return math.exp(t) * math.sin(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 0.30
h = 0.002
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)

Approximate value of y(10) using 4th order Runge-Kutta method: -105.35173254917531


In [103]:
import math

def f(t, y):
    return math.exp(t) * math.sin(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 0.30
h = 0.003
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)

Approximate value of y(10) using 4th order Runge-Kutta method: 161.6033865748213


We see that there is qualitively no change in the numerical approximations being way far from the analytical solution of $\pi$ even after using Runge-Kutta method. This is what step 6 asks for, we are done.

7
____

We want to find similar differential equations where numerical schemes break down. We see that $y'(t)= e^t cos(y)$ with ic $y(0)=1$ also breaks down

In [ ]:
import math

def f(t, y):
    return math.exp(t) * math.cos(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 1
h = 0.30
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)


In [ ]:
import math

def f(t, y):
    return math.exp(t) * math.cos(y)

def runge_kutta_4th_order(t0, y0, h, t_end):
    t = t0
    y = y0
    while t < t_end:
        k1 = h * f(t, y)
        k2 = h * f(t + 0.5*h, y + 0.5*k1)
        k3 = h * f(t + 0.5*h, y + 0.5*k2)
        k4 = h * f(t + h, y + k3)

        y += (k1 + 2*k2 + 2*k3 + k4) / 6.0
        t += h
    return y

# Given values
t0 = 0
y0 = 1
h = 0.301
t_end = 10

y_10 = runge_kutta_4th_order(t0, y0, h, t_end)
print("Approximate value of y(10) using 4th order Runge-Kutta method:", y_10)

So we see that our example for step 7 works.